crates/uv-globfilter/src/glob_dir_filter.rs

```diff
@@ -0,0 +1,277 @@
+        let mut glob_set_builder = GlobSetBuilder::new();
+        for glob in globs {
+            glob_set_builder.add(glob.clone());
+        }
+        let glob_set = glob_set_builder.build()?;
+
+        let regexes: Vec<_> = globs
+            .iter()
+            .map(|glob| {
+                let main_separator = regex::escape(MAIN_SEPARATOR_STR);
+                let regex = glob
+                    .regex()
+                    // We are using a custom DFA builder
+                    .strip_prefix("(?-u)")
+                    .expect("a glob is a non-unicode byte regex")
+                    // Match windows paths if applicable
+                    .replace('/', &main_separator);
+                regex
+            })
+            .collect();
+
+        let dfa_builder = dfa::dense::Builder::new()
+            .syntax(
+                // The glob regex is a byte matcher
+                regex_automata::util::syntax::Config::new()
+                    .unicode(false)
+                    .utf8(false),
+            )
+            .configure(
+                dfa::dense::Config::new()
+                    .start_kind(dfa::StartKind::Anchored)
+                    // DFA can grow exponentially, in which case we bail out
+                    .dfa_size_limit(Some(DFA_SIZE_LIMIT))
+                    .determinize_size_limit(Some(DFA_SIZE_LIMIT)),
+            )
+            .build_many(&regexes);
+        let dfa = if let Ok(dfa) = dfa_builder {
+            Some(dfa)
+        } else {
+            // TODO(konsti): `regex_automata::dfa::dense::BuildError` should allow asking whether
+            // is a size error
+            warn!(
+                "Glob expressions regex is larger than {DFA_SIZE_LIMIT} bytes, \
+                    falling back to full directory traversal!"
+            );
+            None
+        };
+
+        Ok(Self { glob_set, dfa })
+    }
+
+    /// Whether the path (file or directory) matches any of the globs.
+    pub fn match_path(&self, path: &Path) -> bool {
+        self.glob_set.is_match(path)
+    }
+
+    /// Check whether a directory or any of its children can be matched by any of the globs.
+    ///
+    /// This option never returns false if any child matches, but it may return true even if we
+    /// don't end up including any child.
+    pub fn match_directory(&self, path: &Path) -> bool {
+        let Some(dfa) = &self.dfa else {
+            return false;
+        };
+
+        // Allow the root path
+        if path == Path::new("") {
+            return true;
+        }
+
+        let config_anchored =
+            regex_automata::util::start::Config::new().anchored(regex_automata::Anchored::Yes);
+        let mut state = dfa.start_state(&config_anchored).unwrap();
+
+        // Paths aren't necessarily UTF-8, which we can gloss over since the globs match bytes only
+        // anyway.
+        let byte_path = Vec::from_path_lossy(path);
```

Comment:
I believe this is one of the few places where [`OsStr::as_encoded_bytes`](https://doc.rust-lang.org/std/ffi/struct.OsStr.html#method.as_encoded_bytes) is appropriate! This is great because it will avoid an alloc and a UTF-8 validation scan.

Reply:
> 1.74.0

I've waited for this method for years :tada:

Reply:
Been about a decade for me hah.

(Although I've always been skeptical of stabilizing it too, because it makes it much easier to use WTF-8 as an interchange format, which is not supposed to happen.)